### PR TITLE
Add demo to connect spyrit and deepinv

### PR DESCRIPTION
**Pull request by tbaudier**
PR related to https://github.com/deepinv/deepinv/discussions/700#discussioncomment-14243978 with @nducros

Thank you for contributing to DeepInverse!

Please refer to our [contributing guidelines](https://deepinv.github.io/deepinv/contributing.html) for full instructions on how to contribute, including writing tests, documentation and code style.

Once the GitHub tests have been approved by a maintainer (only required for first-time contributors), and the `Build Docs` GitHub action has run successfully, you can download the documentation as a zip file from the [Actions page](https://github.com/deepinv/deepinv/actions/workflows/documentation.yml). Look for the workflow run corresponding to your pull request.

### Checks to be done before submitting your PR

- [ ] `python3 -m pytest deepinv/tests` runs successfully.
- [ ] `black .` runs successfully.
- [ ] `make html` runs successfully (in the `docs/` directory).
- [ ] Updated docstrings related to the changes (as applicable).
- [ ] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

### CODE REVIEW

File: examples/external-libraries/demo_connect_spyrit.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+import torchvision
+import torch.nn
+
+from spyrit.misc.disp import imagesc
+from spyrit.misc.statistics import transform_gray_norm
+
+import deepinv as dinv
+
+spyritPath = os.getcwd()
+imgs_path = os.path.join(spyritPath, "images/")
```
Comment (Tmodrzyk) [inline examples/external-libraries/demo_connect_spyrit.py]: It would be better if we used an image loaded from the internet. This allows both the CI to build this notebook and new users to run the example without worrying about manually adding images. Typically with:
``x = dinv.utils.load_example("butterfly.png", device=device, img_size=(im_size, im_size)``

File: examples/external-libraries/demo_connect_spyrit.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+# Grayscale images of size (32, 32), no normalization to keep values in (0,1)
+transform = transform_gray_norm(img_size=32, normalize=False)
+
+# Create dataset and loader (expects class folder 'images/test/')
+dataset = torchvision.datasets.ImageFolder(root=imgs_path, transform=transform)
+dataloader = torch.utils.data.DataLoader(dataset, batch_size=7)
+
+x, _ = next(iter(dataloader))
+print(f"Ground-truth images: {x.shape}")
```
Comment (Tmodrzyk) [inline examples/external-libraries/demo_connect_spyrit.py]: All of these lines can be remove if we use directly:
`` x = dinv.utils.load_example("butterfly.png", device=device, img_size=(32, 32), grayscale=True``
The grayscale option will normalize between 0 and 1.

File: examples/external-libraries/demo_connect_spyrit.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+###############################################################################
+# We select the second image in the batch and plot it.
+
+i_plot = 1
+imagesc(x[i_plot, 0, :, :], r"$32\times 32$ image $X$")
```
Comment (Tmodrzyk) [inline examples/external-libraries/demo_connect_spyrit.py]: I think it would make more sense to replace this by ``dinv.utils.plot`` (deepinv's plotting function) or to use both of them to compare the displays.

File: examples/external-libraries/demo_connect_spyrit.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+# %%
+# Forward operator
+# ----------------------------------------------------------------------
+
+###############################################################################
+# You can direcly give the forward operator to deepinv. You can also add noise using deepinv model or spyrit model.
+meas_deepinv = dinv.physics.LinearPhysics(
+    lambda y: meas_spyrit.measure_H(y) / norm,
+    A_adjoint=lambda y: meas_spyrit.unvectorize(meas_spyrit.adjoint_H(y) / norm),
+)
```
Comment (Tmodrzyk) [inline examples/external-libraries/demo_connect_spyrit.py]: Here the deepinv physics is defined without Poisson noise, but it seems you wanted to add it in the spyrit physics.
Defining both with Poisson noise and comparing the measures would be great. 
You can add a noise model directy upon declaring the class with:
```
meas_deepinv = dinv.physics.LinearPhysics(
    lambda y: meas_spyrit.measure_H(y) / norm,
    A_adjoint=lambda y: meas_spyrit.unvectorize(meas_spyrit.adjoint_H(y) / norm),
    noise_model=dinv.physics.PoissonNoise(gain=1/50.)
)
```

File: examples/external-libraries/demo_connect_spyrit.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+# meas_deepinv.noise_model = dinv.physics.GaussianNoise(sigma=0.01)
```
Comment (Tmodrzyk) [inline examples/external-libraries/demo_connect_spyrit.py]: I think this line we can remove this line since the noise will be defined in the physics directly.

File: examples/external-libraries/demo_connect_spyrit.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    prior=dinv.optim.TVPrior(),
+    data_fidelity=dinv.optim.L2(),
+    params_algo={"stepsize": 1, "lambda": 5e-2},
+    max_iter=10,
+    custom_init=lambda y, Physics: {"est": (Physics.A_dagger(y),)},
+)
+
+x_tv, metrics_TV = model_tv(m_spyrit / norm, meas_deepinv, compute_metrics=True, x_gt=x)
+dinv.utils.plot_curves(metrics_TV)
+imagesc(x_tv[1, 0, :, :].cpu(), "TV recon")
```
Comment (Tmodrzyk) [inline examples/external-libraries/demo_connect_spyrit.py]: Here we can use ``deepinv.utils.plot`` again, and plot side by side the ground truth, observation, pseudo-inverse and reconstruction. Another suggestion is to add the PSNR values under each of them using the ``subtitles`` option, which we usually do in demos. Something like:
```

metric = dinv.metric.PSNR()

psnr_y = metric(y, x).item()
psnr_pinv = metric(x_pinv, x).item()
psnr_tv = metric(x_tv, x).item()

dinv.utils.plot(
    {
        "Ground Truth": x,
        "Measurement": y,
        "Pseudo-Inverse": x_pinv,
        "Reconstruction": x_tv,
    },
    subtitles=["PSNR:", f"{psnr_y:.2f} dB", f"{psnr_pinv:.2f} dB", f"{psnr_tv:.2f} dB"],
)
```

File: examples/external-libraries/demo_connect_spyrit.py (hunk 2 above)
Comment (nducros) [reply examples/external-libraries/demo_connect_spyrit.py]: OK, to do

File: examples/external-libraries/demo_connect_spyrit.py (hunk 3 above)
Comment (nducros) [reply examples/external-libraries/demo_connect_spyrit.py]: ok, to do

File: examples/external-libraries/demo_connect_spyrit.py (hunk 4 above)
Comment (nducros) [reply examples/external-libraries/demo_connect_spyrit.py]: I would prefer to focus here on the "connection" to spyrit (no noise)

File: examples/external-libraries/demo_connect_spyrit.py (hunk 5 above)
Comment (nducros) [reply examples/external-libraries/demo_connect_spyrit.py]: I would prefer to keep two lines and Gaussian noise (see comment above)

File: examples/external-libraries/demo_connect_spyrit.py (hunk 6 above)
Comment (nducros) [reply examples/external-libraries/demo_connect_spyrit.py]: OK let's try, to do.

File: examples/external-libraries/demo_connect_spyrit.py (hunk 1 above)
Comment (nducros) [reply examples/external-libraries/demo_connect_spyrit.py]: ok, todo.

**Comment by nducros** [comment]
We'll implement your suggestions (see my specific comments). Thanks Thomas!

**Comment by Tmodrzyk** [comment]
Hi @jscanvic could you take a quick look on this example ? After a quick IRL discussion with Nicolas I implemented my own recommendations because it was faster.
A specific point I'm not sure about is that I added ``spyrit`` to the list of optional dependencies, so that the example would build in the doc. I'm not sure if it was the correct solution.

**Comment by tbaudier** [comment]
I have a look on your modification and it's ok for me concerning the spyrit example

**Comment by jscanvic** [comment]
Thanks guys! @tbaudier @nducros @Tmodrzyk I'll have a look shortly

File: examples/external-libraries/demo_connect_spyrit.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
+"""
+Using Spyrit with DeepInverse
+====================================================================================================
+
+This example shows how to use Spyrit linear models and measurements with DeepInverse
+Here we use the HadamSplit2d linear model from Spyrit.
+"""
+
+# %%
+# Loads images
```
Comment (jscanvic) [inline examples/external-libraries/demo_connect_spyrit.py]: For consistency with other pages

```suggestion
# Load images
```

File: examples/external-libraries/demo_connect_spyrit.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
+"""
+Using Spyrit with DeepInverse
+====================================================================================================
+
+This example shows how to use Spyrit linear models and measurements with DeepInverse
```
Comment (jscanvic) [inline examples/external-libraries/demo_connect_spyrit.py]:
```suggestion
This example shows how to use Spyrit linear models and measurements with DeepInverse.
```

(and add a new line if it's intended to be on two lines)

File: examples/external-libraries/demo_connect_spyrit.py [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+device = dinv.utils.get_freer_gpu() if torch.cuda.is_available() else "cpu"
+im_size = 64
+x = dinv.utils.load_example(
+    "butterfly.png", device=device, img_size=(im_size, im_size), grayscale=True
+)
+print(f"Ground-truth images: {x.shape}")
+
+###############################################################################
+# We select the second image in the batch and plot it.
```
Comment (jscanvic) [inline examples/external-libraries/demo_connect_spyrit.py]:
```suggestion
# Then we plot it:
```

File: examples/external-libraries/demo_connect_spyrit.py [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+This example shows how to use Spyrit linear models and measurements with DeepInverse
+Here we use the HadamSplit2d linear model from Spyrit.
+"""
+
+# %%
+# Loads images
+# -----------------------------------------------------------------------------
+
+###############################################################################
+# We first load the butterfly image using deepinv's load_example utility.
```
Comment (jscanvic) [inline examples/external-libraries/demo_connect_spyrit.py]:
```suggestion
# We start by loading the butterfly image using `func`:`deepinv.utils.load_example`:
```

File: examples/external-libraries/demo_connect_spyrit.py [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+
+from deepinv.utils import plot
+import deepinv as dinv
+
+device = dinv.utils.get_freer_gpu() if torch.cuda.is_available() else "cpu"
+im_size = 64
+x = dinv.utils.load_example(
+    "butterfly.png", device=device, img_size=(im_size, im_size), grayscale=True
+)
+print(f"Ground-truth images: {x.shape}")
```
Comment (jscanvic) [inline examples/external-libraries/demo_connect_spyrit.py]:
```suggestion
print(f"Ground-truth image: {x.shape}")
```

File: examples/external-libraries/demo_connect_spyrit.py [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+
+###############################################################################
+# We select the second image in the batch and plot it.
+
+plot(x, r"$32\times 32$ image $X$")
```
Comment (jscanvic) [inline examples/external-libraries/demo_connect_spyrit.py:30]:
```suggestion
plot(x, r"$32\times 32$ image $x$")
```

File: examples/external-libraries/demo_connect_spyrit.py [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+
+# %%
+# Basic example
+# -----------------------------------------------------------------------------
+
+######################################################################
+# We instantiate an HadamSplit2d object and simulate the 2D hadamard transform of the input images. Reshape output is necesary for deepinv. We also add Poisson noise.
```
Comment (jscanvic) [inline examples/external-libraries/demo_connect_spyrit.py:37]:
```suggestion
# We create a `HadamSplit2d` operator from `spyrit` and use it to synthesize measurements from the ground truth:
# 
# .. note::
#    It is necessary to set the option `reshape_output` to ensure compatibility with DeepInverse.
```

(Is Poisson noise added?)

File: examples/external-libraries/demo_connect_spyrit.py [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+######################################################################
+# First, use the adjoint and dagger (pseudo-inverse) operators to reconstruct the image.
+x_adj = physics_deepinv.A_adjoint(y_spyrit / norm)
+
+x_pinv = physics_deepinv.A_dagger(y_spyrit / norm)
+
+######################################################################
+# You can also use optimization-based methods from deepinv. Here, we use Total Variation (TV) regularization with a projected gradient descent (PGD) algorithm. You can note the use of the custom_init parameter to initialize the algorithm with the dagger operator.
```
Comment (jscanvic) [inline examples/external-libraries/demo_connect_spyrit.py:80]:
```suggestion
# Optimization-based solvers work as well:
```

File: examples/external-libraries/demo_connect_spyrit.py [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+
+x_tv, metrics_TV = model_tv(
+    y_spyrit / norm, physics_deepinv, compute_metrics=True, x_gt=x
+)
+
+######################################################################
+# Deep Plug and Play (DPIR) algorithm can also be used with a pretrained denoiser. Here, we use the DRUNet denoiser.
+denoiser = dinv.models.DRUNet(in_channels=1, out_channels=1, device=device)
+model_dpir = dinv.optim.DPIR(sigma=1e-1, device=device, denoiser=denoiser)
+model_dpir.custom_init = lambda y, Physics: {"est": (Physics.A_dagger(y),)}
```
Comment (jscanvic) [inline examples/external-libraries/demo_connect_spyrit.py:98]:
```suggestion
model_dpir.custom_init = lambda y, physics: {"est": (physics.A_dagger(y),)}
```

File: examples/external-libraries/demo_connect_spyrit.py [hunk 16]
```diff
@@ -0,0 +1,132 @@
+"""
+Using Spyrit with DeepInverse
```
Comment (jscanvic) [inline examples/external-libraries/demo_connect_spyrit.py]:
```suggestion
Single-pixel imaging with Spyrit
```
(I'd avoid writing the name of the library too many times)

File: examples/external-libraries/demo_connect_spyrit.py [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+# You can direcly give the forward operator to deepinv. You can also add noise using deepinv model or spyrit model.
+physics_deepinv = dinv.physics.LinearPhysics(
+    lambda y: physics_spyrit.measure_H(y) / norm,
+    A_adjoint=lambda y: physics_spyrit.unvectorize(physics_spyrit.adjoint_H(y) / norm),
+)
+y_deepinv = physics_deepinv(x)
+print("diff:", torch.linalg.norm(y_spyrit / norm - y_deepinv))
+
+# %%
+# Reconstruction with deepinverse
```
Comment (jscanvic) [inline examples/external-libraries/demo_connect_spyrit.py]:
```suggestion
# Computing the reconstructions
```

File: examples/external-libraries/demo_connect_spyrit.py [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+######################################################################
+# First, use the adjoint and dagger (pseudo-inverse) operators to reconstruct the image.
```
Comment (jscanvic) [inline examples/external-libraries/demo_connect_spyrit.py]:
```suggestion
# All of the usual solvers work out of the box and we showcase some of them here starting with simple linear reconstructions using :meth:`deepinv.physics.LinearPhysics.A_adjoint` and :meth:`deepinv.physics.LinearPhysics.A_dagger`:
```

File: examples/external-libraries/demo_connect_spyrit.py [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+    max_iter=10,
+    custom_init=lambda y, Physics: {"est": (Physics.A_dagger(y),)},
+)
+
+x_tv, metrics_TV = model_tv(
+    y_spyrit / norm, physics_deepinv, compute_metrics=True, x_gt=x
+)
+
+######################################################################
+# Deep Plug and Play (DPIR) algorithm can also be used with a pretrained denoiser. Here, we use the DRUNet denoiser.
```
Comment (jscanvic) [inline examples/external-libraries/demo_connect_spyrit.py]:
```suggestion
# And so do deep learning methods:
```

File: examples/external-libraries/demo_connect_spyrit.py [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+denoiser = dinv.models.DRUNet(in_channels=1, out_channels=1, device=device)
+model_dpir = dinv.optim.DPIR(sigma=1e-1, device=device, denoiser=denoiser)
+model_dpir.custom_init = lambda y, Physics: {"est": (Physics.A_dagger(y),)}
+with torch.no_grad():
+    x_dpir = model_dpir(y_spyrit / norm, physics_deepinv)
+
+######################################################################
+# Reconstruct Anything Model (RAM) can also be used.
```
Comment (jscanvic) [inline examples/external-libraries/demo_connect_spyrit.py]:
```suggestion
# Including reconstruction with :class:`deepinv.models.RAM`:
```

**Review by Tmodrzyk — APPROVED** [review]
LGTM, thanks @jscanvic for improving the notebook! :smiley:

**Review by jscanvic — APPROVED** [review]
Thanks @Tmodrzyk!
